### ISSUE TITLE
Feature request: Allow to determine charged payment fee for a specific payment

### ISSUE THREAD

**Issue by holtkamp** [mollie/mollie-api-php#296]
It would be nice to be able to determine the payment fee / transaction cost that was charged for a _specific_ Payment X.

It currently IS possible to acquire the total costs of processed Payments using a certain method, but for transaction costs which involve a percentual part, it would be nice to be able to determine the exact charges of a _particular_ Payment X.

**Observations**
When using the various API endpoints:

- The Payment X object does not contain this information
- The Settlement Y object that "covers" Payment X only contains the cumulative costs of all Payments in the Settlement, for example:
``` 
["costs"] => array(2) {
        [0] => object(stdClass)#23225 (5) {
          ["description"] => string(5) "iDEAL"
          ["method"] => string(5) "ideal"
          ["count"] => int(44)
          ["rate"] => object(stdClass)#23226 (2) {
            ["fixed"] => string(6) "0.2900"
            ["percentage"] => NULL
          }
          ["amount"] => object(stdClass)#23227 (3) {
            ["net"] => string(7) "12.7600"
            ["vat"] => string(6) "2.6796"
            ["gross"] => string(7) "15.4396"
          }
        }
        [1] => object(stdClass)#23228 (5) {
          ["description"] => string(6) "PayPal"
          ["method"] => string(6) "paypal"
          ["count"] => int(1)
          ["rate"] => object(stdClass)#23229 (2) {
            ["fixed"] => string(6) "0.1000"
            ["percentage"] => NULL
          }
          ["amount"] => object(stdClass)#23230 (3) {
            ["net"] => string(6) "0.1000"
            ["vat"] => string(6) "0.0210"
            ["gross"] => string(6) "0.1210"
          }
        }
      }

```
- The Invoice Z object that "covers" Settlement Y (also) only contains the cumulative costs of all involved Payments in the same format as the snippet above.

Especially for payments that involve percentual charges, the *actually* charged costs are required for administration and invoicing purposes. This particularly applies for payment methods with "different" percentages, like the MasterCard and Visa CreditCards:

<img width="940" alt="screenshot 2018-10-10 at 14 51 22" src="https://user-images.githubusercontent.com/776405/46737545-125f4c80-cc9c-11e8-81b4-67a04e29a3e8.png">

*Disclaimer*
Maybe the information can already be extracted from the API, could not find it though...

**Comment by holtkamp** [comment]
Hey guys, just wondering: is this a "valid" feature request? Can it be confirmed as such (or bounced of course 🤓 )

**Comment by mollierobbert** [comment]
For sure, Menno. Thank you for taking the time to write this proposal. We are working actively on upgrading our engine to be able to provide real-time payment fee insights in the near future. We cannot confirm a date yet, however, hence we can't promise much yet.

**Comment by sandervanhooft** [comment]
@holtkamp not entirely the same, but does the latest release solve your use case? 

https://github.com/mollie/mollie-api-php/releases/tag/v2.5.0

**Comment by holtkamp** [comment]
@sandervanhooft thanks for the heads-up. 

>but does the latest release solve your use case?

Mmm, I haven't looked at in detail, but it seems the `pricing()` helper gives information about the pricing of a Payment method. Not the actually charged / to-be-invoiced costs of an actually processed transaction.

**Comment by holtkamp** [comment]
>We are working actively on upgrading our engine to be able to provide real-time payment fee insights in the near future

@mollierobbert just wondering, are there any updates regarding this issue? 


**Rationale**
Currently, under PSD2 "surcharging" is still allowed to some extent. Also see:

- https://help.mollie.com/hc/nl/articles/360012564454-Mag-ik-de-kosten-die-ik-maak-voor-betaalmethodes-doorberekenen-aan-mijn-klanten-
- https://www.acm.nl/nl/publicaties/invoering-psd2-wat-gaat-er-voor-u-veranderen

However, I expect that surcharging will (eventually) become completely unallowed. In that case the situation will occur that the charged transaction costs need to be forwarded/invoiced elsewhere. To generate such invoices, it would be nice to have insight in the actually charged transaction fees. 

This does not need to be "real-time", but at the moment a Mollie invoice is received (by the end of the month), it would be nice to have insight in these details...

**Comment by mollierobbert** [comment]
@holtkamp The pricing is currently determined on a settlement level. You can find the fees by retrieving the settlement via the Settlements API (see `periods`):

* https://docs.mollie.com/reference/v2/settlements-api/get-settlement

Alternatively, you can wait for the monthly invoice to be generated and fetch the fees from the invoice using the Invoices API (see `lines`):

* https://docs.mollie.com/reference/v2/invoices-api/get-invoice

We are indeed still working on calculating fees in real-time, but I cannot share a timeline yet for when that feature would reach our Payments API.

**Comment by MollieRick** [comment]
@holtkamp Is it okay for you to close this issue for now?

**Comment by christopheg** [comment]
Dear @mollierobbert, can you please explain why this is so difficult and why it takes so long to implement?
Basically we could do the same by following this procedure:
- On webhook callback, check if the status is 'paid'
- fetch the fees via the 'methods' API call
- try to match the brand/cardAudience/feeRegion with fetched fees
- do the calculation

Problems that may arise are situated in:
- matching the fees with the chosen method: There is no ways to link the response from the 'method' call with the response of a payment/get call. This will end up in a very big if/then/else statement that needs to be updated on every adjustment on the side of Mollie. 
- The calculation: We will probably end up with rounding errors. The goal should be that we could match the totals mentioned on the settlements.

These problems are easily overcome if Mollie would implement this procedure on there side.

**Comment by mollierobbert** [comment]
Thanks for the suggestion @christopheg.

We currently can only provide an overview of the fees once the settlement has been processed. Would it suffice for you both, @christopheg and @holtkamp, if these fees are made available in the Payments API after the payment got settled?

As noted before we are actively working on a big engine upgrade that will allow us to do provide this information as soon as the payment moves to 'paid'. We are getting there, slowly but surely.

**Comment by holtkamp** [comment]
>Would it suffice for you both, @christopheg and @holtkamp, if these fees are made available in the Payments API after the payment got settled (i.e. to your bank account)?

That would be a nice first step indeed 😃

**Comment by christopheg** [comment]
@mollierobbert, in our use-case this is not sufficient. We charge our customers the exact cost of the transaction.
I just don't understand why you need a bit engine upgrade in order to create something which we could develop ourselves with existing API calls...

**Comment by christopheg** [comment]
@mollierobbert  I've created a pull request that adds cost calculation of transactions via existing API calls. This is for me a temporary solution but it fixes our problem. 
Pull request: https://github.com/mollie/mollie-api-php/pull/409

PS: I want to point out that this is not only a technical limitation of the Mollie system. For the moment there is no way to clarify the costs that are calculated on any invoice. I think that therefore also from a commercial point-of-view it is important to come up with a better solution.

**Comment by mollierobbert** [comment]
Thanks for the pull request, Christophe. 👏

We could indeed take your suggested shortcut, but we would not be able to expose this endpoint to our entire user base reliably, since costs are simply currently not yet confirmed in real-time. It would solve your problem, and cause issues for us elsewhere (for users with more complex pricing models).

I do of course agree on the importance of this issue, both from a business and a technical standpoint.

I'm tempted to merge the pull request, but we would have to put heavy disclaimers around it that it's highly experimental. @willemstuursma @leonardogcsoares what do you guys think?

Alternatively, you are free to fork the package.

@holtkamp Would Christophe's PR solve your issue for now as well?

**Comment by sandervanhooft** [comment]
In order to prevent breaking changes down the road, I strongly recommend to only use this code locally. As soon as the API is ready this code is very likely to be deprecated.

If this is the feature set you’re looking for and you are adamant to have a solution via the client class I suggest to decorate it locally for now.

**Comment by sandervanhooft** [comment]
“Locally” being “in your application”.

**Comment by holtkamp** [comment]
@mollierobbert 

>Would Christophe's PR solve your issue for now as well?

Well, yes. But I understand the arguments to not adopt this (approach) in the API and advise to keep it in "user land" code instead.

**Comment by christopheg** [comment]
I also agree on the fact that it is not ideal to have this code in the Mollie API client. However I was looking for a way to distribute the code in such a way that in can be improved and maintained by you or other developers. Do you see any other solution?

**Comment by sandervanhooft** [comment]
Good thinking @christopheg , I'd recommend a dedicated package for that, pulling in this client. Once the feature set is available on the client itself, you could deprecate the package or refactor it to use the then-new client features instead.

**Comment by sandervanhooft** [comment]
If that's a bit "too much", consider a github gist.

**Comment by mollierobbert** [comment]
A gist would be nice. I was thinking we could also just leave the PR open until we can offer the functionality natively, but that's a bit messy perhaps.

Thanks all for collaborating on this. 👍 And for understanding our delay in solving this issue.

**Comment by noezdev** [comment]
Are there any new developments on the matter? Especially when receiving callbacks via credit card payments, it is almost impossible to calculate the correct fees.

This is mandatory for our accounting. The Mollie support always refers to the settlements and that you have to calculate it manually, but I think this is extremely inconvenient.

Especially since the fees per payment are also displayed in the Mollie backend under transactions, I assume a technical possibility.

**Comment by mollierobbert** [comment]
Thanks for the reply noezdev, and sorry for the wait on this topic.

We are planning to roll out a more mature reporting API in the next 6 months or so. It took some time for us to migrate everyone to a new balancing system that enables us to build more reliable reporting for our users.

The new reporting APIs will likely offer:

* Aggregations of all received payments and charged fees within a given period (e.g. 5 iDEAL payments totaling €50, 5 iDEAL fees totaling €1,45).
* A list of all transactions happening on your balance (e.g. iDEAL payment of €10 with a €0,29 total fee and €9,71 settled on your balance).

It doesn't (yet) offer a breakdown of what fees exactly were charged on each payment, but it does offer the fee sum per payment, which is a great leap forward.

The functionality is already available in the dashboard, so the next step is to 'soften the edges' and publish it as a public API.

I sadly cannot offer an inbetween solution, but if you are interested I can put you on the list for beta testing the new reporting endpoints once they are ready to go into beta.

**Comment by Jay031** [comment]
@mollierobbert any update on this?
I still use this quick fix, but it does't support directdebit and some other methods.
So 50% of my transactions don't get any transactionCosts.

**Comment by christopheg** [comment]
Hi @Jay031,
Unfortunatly, I don't think Mollie wants to implement this feature so don't expect a solution by them....
I created the PR but on my account not all payment methods are enabled. I believe that my patch can be extended for other methods. Just try to list all your payment methods via API and check the 'id'. With this 'id', the PR can be extended.

**Comment by Jay031** [comment]
> Hi @Jay031, Unfortunatly, I don't think Mollie wants to implement this feature so don't expect a solution by them.... I created the PR but on my account not all payment methods are enabled. I believe that my patch can be extended for other methods. Just try to list all your payment methods via API and check the 'id'. With this 'id', the PR can be extended.

Yeah well, the annoying thing is, that my recurring payments only show 'directdebit' as payment method, so no iDeal, KBC/CBC, Sofort etc. thats only in the initial payment. So that leaves an annoying amount of extra work, I think it's so lame, because the data is already there in Mollie..

**Comment by mollierobbert** [comment]
No update yet, sadly. We are getting closer to a beta release, but my earlier time indication was clearly too optimistic.

Before we can put the endpoints in beta, we still have to address some performance issues and add multicurrency compatibility.

**Comment by RobinvanderVliet** [comment]
Is there an update in the meantime? In the API documentation of the payment endpoints, I still cannot find any information about the charged payment fees.

**Comment by mollierobbert** [comment]
The new Balances API is currently in open beta, available here:

https://docs.mollie.com/reference/v2/balances-api/overview

With this API you can fetch lists of balance transactions. A payment is one type of balance transaction, and fee information is included in the transaction details as 'deductions'.

Perhaps more interesting is the balance reporting endpoint though, which allows you to fetch aggregate reports for a given time period. The aggregate report will include aggregated information on how many payments came in, how many fees were charged, etc.

Hope this helps!

**Comment by sandervanhooft** [comment]
Hi @owebia ,

These can really be best obtained from the Balance endpoint. You don't need a full Mollie Connect OAuth integration here, an organization token would do, which you can create in the Mollie dashboard.

**Comment by owebia** [comment]
Hi @sandervanhooft ,

Thanks for your answer.  
I made some tests but I'm not able to retrieve anything from the balance transactions endpoint.

Here is my code:

```php
$result = $this->client->balances->primary();
echo json_encode($result, JSON_PRETTY_PRINT);
```

✅ Result:
```json
{
    "resource": "balance",
    "mode": "live",
    "id": "bal_XXX",
    ...
    "pendingAmount": {
        "value": "96.85",
        "currency": "EUR"
    }
}
```

---

```php
$result = $this->client->balanceTransactions->pageForPrimary();
echo json_encode($result, JSON_PRETTY_PRINT);
```

❌ Result:
```json
{}
```

---

```php
$result = $this->client->balanceReports->getForPrimary(
    'from' => (new \DateTime())->modify('-30 days')->format('Y-m-d'),
    'until' => (new \DateTime())->format('Y-m-d'),
);
echo json_encode($result, JSON_PRETTY_PRINT);
```

✅ Result:
```json
{
    "resource": "balance-report",
    "balanceId": "bal_XXX",
    "timeZone": "Europe\/Amsterdam",
    "from": "2025-11-12",
    "until": "2025-12-12",
    "grouping": "status-balances",
    "totals": {
        ...
    },
    "_links": {
        ...
    }
}
```

Does the balance transactions endpoint only return data after the first transfer or after the first payment on a balance?
Anything else I missed?

I need to be able to match the corresponding fees of a payment by its ID `tr_...`

**Comment by sandervanhooft** [comment]
> I need to be able to match the corresponding fees of a payment by its ID `tr_...`

The balance transactions endpoint is the way to go here. You can use the payment id from the context to match your records.

I am not entirely sure when these transactions become accessible,  I'd expect these to be available even when the balance itself still is pending/open.

Example balance transactions response:

```json
{
  "count": 2,
  "_embedded": {
    "balance_transactions": [
      {
        "resource": "balance-transaction",
        "id": "baltr_QM24QwzUWR4ev4Xfgyt29A",
        "type": "refund",
        "resultAmount": {
          "currency": "EUR",
          "value": "-10.25"
        },
        "initialAmount": {
          "currency": "EUR",
          "value": "-10.00"
        },
        "deductions": {
          "currency": "EUR",
          "value": "-0.25"
        },
        "createdAt": "2022-12-16T11:23:25+00:00",
        "context": {
          "paymentId": "tr_5B8cwPMGnU6qLbRvo7qEZo",
          "refundId": "re_4qqhO89gsT"
        }
      },
      {
        "resource": "balance-transaction",
        "id": "baltr_WhmDwNYR87FPDbiwBhUXCh",
        "type": "payment", // <-- FILTER ON THIS LOCALLY
        "resultAmount": {
          "currency": "EUR",
          "value": "9.71"
        },
        "initialAmount": {
          "currency": "EUR",
          "value": "10.00"
        },
        "deductions": {
          "currency": "EUR",
          "value": "-0.29"
        },
        "createdAt": "2022-12-16T11:23:25+00:00",
        "context": {
          "paymentId": "tr_5B8cwPMGnU6qLbRvo7qEZo" // <-- USE THIS TO MAP YOUR RECORDS
        }
      }
    ]
  },
  "_links": {
    "self": {
      "href": "...",
      "type": "application/hal+json"
    },
    "previous": null,
    "next": {
      "href": "https://api.mollie.com/v2/balances/bal_gVMhHKqSSRYJyPsuoPNFH/transactions?from=baltr_rXeW2yPqqDUyfAqq8fS5Bg&limit=5",
      "type": "application/hal+json"
    },
    "documentation": {
      "href": "...",
      "type": "text/html"
    }
  }
}
```

**Comment by erik-re** [comment]
Hi,

using this endpoint:
https://api.mollie.com/v2/balances/bal_xxx/transactions?limit=250
i receive a large list of items, but unfortunately no creditcard transactions are in this list (i get entries for all PayPal transactions which i also see in the mollie backend ui, but every cc row in mollie backend is missing in the transactions-endpoint).

Am i missing a parameter? I need to fetch the fees for creditcard-transactions.
